### PR TITLE
fix(gfql): handle schema-changing operations in chains

### LINKED ISSUE #761: UMAP in GFQL chains fails with missing tracking column 'index'
# PyGraphistry Bug Report: UMAP in GFQL Chains Fails with Missing Tracking Column

**Repository**: graphistry/pygraphistry
**Version**: v0.43.0
**Component**: `graphistry/compute/chain.py:191` in `combine_steps()`

## Summary

UMAP operations fail when used in GFQL chains (even simple 1-operation chains) with error:
```
ValueError: Column 'index' not found in edges step DataFrame.
Step has id='index', available columns: ['_src_implicit', '_dst_implicit', '_weight'].
```

## Minimal Reproduction

```python
import pandas as pd
import graphistry

# Create simple graph
edges_df = pd.DataFrame({
    'src': ['A', 'B', 'C', 'D', 'E'],
    'dst': ['B', 'C', 'D', 'E', 'A']
})

nodes_df = pd.DataFrame({
    'id': ['A', 'B', 'C', 'D', 'E'],
    'town': ['Birchford', 'Cedarland', 'Maplewood', 'Oakville', 'Pineville'],
    'poi': ['store', 'theatre', 'pet_shop', 'square', 'park']
})

# Create plottable
g = graphistry.edges(edges_df, 'src', 'dst').nodes(nodes_df, 'id')

# Try UMAP in a GFQL chain - THIS FAILS
from graphistry.compute.ast import ASTCall

umap_op = ASTCall('umap', {
    'n_components': 2,
    'n_neighbors': 3,
    'umap_kwargs': {'random_state': 42}
})

# This will fail even with just one operation
result = g.gfql([umap_op], engine='pandas')
```

## Error Output

```
ValueError: Column 'index' not found in edges step DataFrame.
Step has id='index', available columns: ['_src_implicit', '_dst_implicit', '_weight'].
Operation: <graphistry.compute.ast.ASTCall object at 0x...>

Traceback:
  File "graphistry/compute/gfql_unified.py", line 257, in gfql
    return chain_impl(self, converted_query, engine, policy=policy)
  File "graphistry/compute/chain.py", line 275, in chain
    return _chain_impl(self, ops, engine, validate_schema, policy)
  File "graphistry/compute/chain.py", line 512, in _chain_impl
    final_edges_df = combine_steps(g, 'edges', list(zip(ops, reversed(g_stack_reverse))), engine_concrete)
  File "graphistry/compute/chain.py", line 191, in combine_steps
    raise ValueError(f"Column '{id}' not found in {kind} step DataFrame. ...")
```

## Root Cause

1. PyGraphistry's chain implementation adds tracking columns (like 'index') to DataFrames to track which operation produced each row
2. UMAP creates completely new edges with only `['_src_implicit', '_dst_implicit', '_weight']` columns
3. UMAP rebinds the graph with these new edges (which is correct behavior)
4. When `combine_steps()` tries to merge results, it looks for the tracking column 'index' but UMAP's new edges don't have it

## Expected Behavior

UMAP should work in GFQL chains. Since UMAP creates an entirely new graph with new edges, the chain code should recognize this and not try to combine it with tracking columns from previous steps.

## Actual Behavior

Chain fails because `combine_steps()` assumes all operations preserve DataFrame structure and tracking columns.

## Suggested Fix

Option 1: **Skip tracking for graph-creating operations**
- Detect operations like UMAP that create new graphs (by checking if edges/nodes are completely replaced)
- Don't try to combine them using tracking columns

Option 2: **Make tracking columns optional in combine_steps**
- Handle missing tracking columns gracefully
- If tracking column is missing, assume the operation created a new graph

Option 3: **UMAP preserves tracking columns**
- Have UMAP copy tracking columns from input to output
- However, this doesn't make semantic sense since UMAP's edges are completely different from the input

## Impact

- UMAP cannot be used in any GFQL chain (even single-operation chains that go through the chain code path)
- Blocks workflows that need UMAP + other operations (e.g., UMAP followed by filtering or degree calculations)

## Workaround

None. UMAP can only be used via direct API calls like `g.umap()`, not through GFQL chains.

## Environment

- PyGraphistry version: v0.43.0
- Python version: 3.10
- Engine tested: both 'pandas' and 'cudf'
- Context: Discovered during GFQL remote operations testing in Graphistry server v2.44.0-rc2

## Additional Notes

The issue specifically occurs in `combine_steps()` which is called by the chain implementation. The error message shows it's looking for a column with `id='index'` which is a tracking column added by the chain machinery, not something from the user's data.

UMAP correctly creates new edges with new bindings - the bug is that chain doesn't handle this case properly.

### PR DESCRIPTION
**Pull request by lmeyerov**
Fixes #761

## Summary

Fix "Column 'index' not found in edges" error when UMAP or hypergraph operations are mixed with other GFQL operations in chains.

Implemented recursive dispatch that automatically splits chains at schema-changer boundaries, executing them as: `before → schema_changer → rest`. This ensures schema-changing operations (which create entirely new graph structures) are properly isolated from regular chain machinery.

### Added
- **UMAP in chains**: Can now use `call('umap', {...})` mixed with filters and other operations
- **Hypergraph in chains**: Removed mixing restriction - now allows `[n(...), call('hypergraph', {...})]`

### Implementation
- `graphistry/compute/chain.py` (lines 372-406): Recursive dispatch logic
  - Detects schema-changers (`['umap', 'hypergraph']`) in operation list
  - Handles singleton, multi-op, and nested cases
  - Falls through to existing chain logic when no schema-changers present
- `graphistry/tests/compute/test_gfql_schema_changers.py`: New comprehensive test suite (8 scenarios)
- `graphistry/tests/compute/test_gfql_hypergraph.py`: Updated to reflect new behavior

#### Unit Tests
✅ Unit tests (compute): 396 passed, 49 skipped  
✅ GFQL tests: 17/17 pass  
✅ Chain tests: 25 passed, 3 skipped  
✅ Lint (flake8): Pass  
✅ Typecheck (mypy): Pass  

#### Standalone Validation (10 scenarios in `plans/fix-umap-chain-tracking/`)
✅ **Singleton operations**: UMAP alone, hypergraph alone  
✅ **Multiple schema-changers**: UMAP→UMAP, UMAP→hypergraph, n→umap→e→hypergraph→n  
✅ **Schema-changers at different positions**: First, middle, last  
✅ **Mixed with filters**: Filter→UMAP, UMAP→filter  
✅ **Edge cases**: Empty before/after segments, consecutive schema-changers  
✅ **Let bindings**: Schema-changers nested in `let()` constructs work correctly

All 10 standalone tests pass ✅

## Example Usage

```python
# Before: This would fail with "Column 'index' not found"
result = g.gfql([
    n({'type': 'person'}),
    call('umap', {'n_neighbors': 15}),
    e()
])

# Now: Works! Chain is split and executed as:
# g.chain([n(...)]) → g.chain([umap(...)]) → g.chain([e()])
```

## Testing Notes

Schema-changers work correctly in:
1. **Direct chains**: `[n(), umap(), e()]` 
2. **Let bindings**: `let({'embedded': call('umap', ...)})`
3. **Multiple schema-changers**: `[umap(), hypergraph()]`
4. **Any position**: Beginning, middle, or end of chain

### CODE REVIEW

File: graphistry/compute/chain.py [hunk 1]
```diff
@@ -369,32 +369,41 @@ def _chain_impl(self: Plottable, ops: Union[List[ASTObject], Chain], engine: Uni
...
-            # We know ops[0] is an ASTCall with hypergraph function
-            hypergraph_call = ops[0]  # This is guaranteed to be ASTCall from the filter above
-            if isinstance(hypergraph_call, ASTCall):
-                return execute_call(self, 'hypergraph', hypergraph_call.params, engine_concrete, policy=policy)
-            else:
-                raise TypeError(f"Expected ASTCall but got {type(hypergraph_call)}")
+            schema_changer = ops[0]
+            # Type narrowing: we know it's ASTCall from the isinstance check above
+            assert isinstance(schema_changer, ASTCall), "Expected ASTCall for schema-changer"
+            return execute_call(self, schema_changer.function, schema_changer.params, engine_concrete, policy=policy)
```
Comment (lmeyerov) [inline graphistry/compute/chain.py:410]: propgate validate_schema?

File: graphistry/compute/chain.py [hunk 2]
```diff
@@ -369,32 +369,41 @@ def _chain_impl(self: Plottable, ops: Union[List[ASTObject], Chain], engine: Uni
...
             engine_concrete = resolve_engine(engine, self)
-            # We know ops[0] is an ASTCall with hypergraph function
-            hypergraph_call = ops[0]  # This is guaranteed to be ASTCall from the filter above
-            if isinstance(hypergraph_call, ASTCall):
-                return execute_call(self, 'hypergraph', hypergraph_call.params, engine_concrete, policy=policy)
-            else:
-                raise TypeError(f"Expected ASTCall but got {type(hypergraph_call)}")
+            schema_changer = ops[0]
+            # Type narrowing: we know it's ASTCall from the isinstance check above
+            assert isinstance(schema_changer, ASTCall), "Expected ASTCall for schema-changer"
```
Comment (lmeyerov) [inline graphistry/compute/chain.py]: structured exn?
